src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -320,37 +386,92 @@ private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBuffe
+        private async Task<Solution> FixSolutionAsync(Solution solution, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            progressTracker.AddItems(solution.ProjectIds.Count);
+            foreach (var projectId in solution.ProjectIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
 
-                    var isRemoveUnusedUsingsEnabled = context.EnabledFixIds.IsFixIdEnabled(RemoveUnusedImportsFixId);
-                    var isSortUsingsEnabled = context.EnabledFixIds.IsFixIdEnabled(SortImportsFixId);
-                    var enabledDiagnostics = new EnabledDiagnosticOptions(enabedDiagnosticSets.ToImmutableArray(),
-                        new OrganizeUsingsSet(isRemoveUnusedUsingsEnabled, isSortUsingsEnabled));
+                var project = solution.GetProject(projectId);
+                progressTracker.Description = project.Name;
 
-                    var newDoc = await codeCleanupService.CleanupAsync(
-                        document, enabledDiagnostics, progressTracker, cancellationToken).ConfigureAwait(true);
+                // FixProjectAsync reports progress for documents within a project, but we limit progress reporting for
+                // a solution to the current project.
+                var projectProgressTracker = new ProgressTracker();
 
-                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-                    return document.Project.Solution.Workspace.TryApplyChanges(newDoc.Project.Solution, progressTracker);
+                var newProject = await FixProjectAsync(project, enabledFixIds, projectProgressTracker, cancellationToken).ConfigureAwait(false);
+                solution = newProject.Solution;
+                progressTracker.ItemCompleted();
+            }
+
+            return solution;
+        }
+
+        private async Task<Project> FixProjectAsync(Project project, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            if (project.LanguageServices.GetService<ICodeCleanupService>() == null)
+            {
+                return project;
+            }
+
+            progressTracker.AddItems(project.DocumentIds.Count);
+            foreach (var documentId in project.DocumentIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var document = project.GetDocument(documentId);
+                progressTracker.Description = document.Name;
+
+                // FixDocumentAsync reports progress within a document, but we limit progress reporting for a project
+                // to the current document.
+                var documentProgressTracker = new ProgressTracker();
+
+                var fixedDocument = await FixDocumentAsync(document, enabledFixIds, documentProgressTracker, cancellationToken).ConfigureAwait(false);
```

Comment:
>FixDocumentAsync [](start = 42, length = 16)

Are we assuming the fix scope is limited to original document? Looking at the list of exported fix IDs, it seems make field readonly is the only current one whose analysis scope is bigger then a single document.

Reply:
It assumes that _changes_ only occur in the original document. Even though the analysis scope of Make Field Readonly is larger than a single document, the changes it applies are restricted to the file in which the diagnostic is reported.